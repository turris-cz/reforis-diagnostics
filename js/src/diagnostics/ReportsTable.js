/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import { Button, DownloadButton, SpinnerElement } from "foris";

import API_URLs from "API";
import { useDeleteReport, useReportIsReady } from "./hooks";

ReportsTable.propTypes = {
    reports: PropTypes.array.isRequired,
    onReload: PropTypes.func.isRequired,
};

export default function ReportsTable({ reports, onReload }) {
    if (reports.length === 0) {
        return (
            <p className="text-muted text-center">
                {_("There are no reports available.")}
            </p>
        );
    }

    return (
        <table className="table table-hover" data-testid="reports-table">
            <tbody>
                <tr>
                    <th scope="col">{_("ID")}</th>
                    <th scope="col" aria-label={_("Actions")} />
                </tr>
                {reports.map((report) => (
                    <ReportRow
                        key={report.diag_id}
                        report={report}
                        onReload={onReload}
                    />
                ))}
            </tbody>
        </table>
    );
}

ReportRow.propTypes = {
    report: PropTypes.object.isRequired,
    onReload: PropTypes.func.isRequired,
};

function ReportRow({ report, onReload }) {
    return (
        <tr>
            <td className="align-middle">{report.diag_id}</td>

            <td className="text-center">
                <ReportActions report={report} onReload={onReload} />
            </td>
        </tr>
    );
}

ReportActions.propTypes = {
    report: PropTypes.shape({
        diag_id: PropTypes.string.isRequired,
    }),
    onReload: PropTypes.func.isRequired,
};

function ReportActions({ report, onReload }) {
    const isReady = useReportIsReady(report);
    const deleteReport = useDeleteReport(report.diag_id, onReload);

    if (!isReady) {
        return <SpinnerElement />;
    }

    return (
        <div className="btn-group" role="group">
            <DownloadButton
                href={`${API_URLs.reports}/${report.diag_id}/contents`}
            >
                {_("Download")}
            </DownloadButton>
            <Button onClick={deleteReport} className="btn-danger btn-sm">
                {_("Delete")}
            </Button>
        </div>
    );
}
