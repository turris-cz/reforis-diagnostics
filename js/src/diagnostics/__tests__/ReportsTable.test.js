/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, getByText, fireEvent, wait } from "customTestRender";
import mockAxios from 'jest-mock-axios';

import { AlertContext } from "foris";

import ReportsTable from "../ReportsTable";

describe("<ReportsTable />", () => {
    const handleReload = jest.fn(),
        setAlert = jest.fn();

    function createTable(reports) {
        const { container } = render(
            <AlertContext.Provider value={setAlert}>
                <ReportsTable reports={reports} onReload={handleReload} />
            </AlertContext.Provider>
        );
        return container;
    }

    it("should render reports", async () => {
        const container = createTable([{diag_id: 1234, status: "ready"}])
        expect(container).toMatchSnapshot();
    });

    it("should handle lack of reports", async () => {
        const container = createTable([])
        expect(container).toMatchSnapshot();
    });

    it("should display spinner when report is not ready", async () => {
        const container = createTable([{diag_id: 1234, status: "pending"}])
        expect(container).toMatchSnapshot();

        // Intial check for status
        expect(mockAxios.get).toBeCalledWith("/reforis/diagnostics/api/reports/1234", expect.anything());
        mockAxios.mockResponse({data: {diag_id: 1234, status: "pending"}});
        // Repeated check for status
        await wait(
            () => expect(mockAxios.get).nthCalledWith(2, "/reforis/diagnostics/api/reports/1234", expect.anything())
        );
        mockAxios.mockResponse({data: {diag_id: 1234, status: "ready"}});
        await wait(() => expect(handleReload).toBeCalledTimes(1));
    });

    it("should handle error on fetching reports", async () => {
        createTable([{diag_id: 1234, status: "pending"}])
        mockAxios.mockError({response: {}});
        await wait(() => {
            expect(setAlert).toHaveBeenCalledWith("Cannot fetch report data");
        });
    });

    it("should delete report", async () => {
        const container = createTable([{diag_id: 1234, status: "ready"}])
        fireEvent.click(getByText(container, "Delete"));
        expect(mockAxios.delete).toHaveBeenCalledWith("/reforis/diagnostics/api/reports/1234", expect.anything());
        // Response to DELETE report
        mockAxios.mockResponse({});
        await wait(() => expect(handleReload).toBeCalledTimes(1));
    });

    it("should handle delete error", async () => {
        const container = createTable([{diag_id: 1234, status: "ready"}])
        fireEvent.click(getByText(container, "Delete"));
        // Response to DELETE report
        mockAxios.mockError({response: {headers: {"content-type": "application/json"}}});
        await wait(() => {
            expect(setAlert).toHaveBeenCalledWith("Cannot delete report");
        });
    });
});
