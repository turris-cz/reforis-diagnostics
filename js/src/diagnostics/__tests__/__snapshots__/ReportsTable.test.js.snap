// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportsTable /> should display spinner when report is not ready 1`] = `
<div>
  <table
    class="table table-hover"
    data-testid="reports-table"
  >
    <tbody>
      <tr>
        <th
          scope="col"
        >
          ID
        </th>
        <th
          aria-label="Download"
          scope="col"
        />
        <th
          aria-label="Delete"
          scope="col"
        />
      </tr>
      <tr>
        <td
          class="align-middle"
        >
          1234
        </td>
        <td>
          <div
            class="spinner-border "
            role="status"
          >
            <span
              class="sr-only"
            />
          </div>
          <div
            class="spinner-text"
          />
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`<ReportsTable /> should handle lack of reports 1`] = `
<div>
  <p
    class="text-muted text-center"
  >
    There are no reports available.
  </p>
</div>
`;

exports[`<ReportsTable /> should render reports 1`] = `
<div>
  <table
    class="table table-hover"
    data-testid="reports-table"
  >
    <tbody>
      <tr>
        <th
          scope="col"
        >
          ID
        </th>
        <th
          aria-label="Download"
          scope="col"
        />
        <th
          aria-label="Delete"
          scope="col"
        />
      </tr>
      <tr>
        <td
          class="align-middle"
        >
          1234
        </td>
        <td>
          <a
            class="btn btn-primary"
            href="/reforis/diagnostics/api/reports/1234/contents"
          >
            Download
          </a>
        </td>
        <td>
          <button
            class="btn btn-danger"
            type="button"
          >
             
             
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
